builder/builder.go

```diff
@@ -0,0 +1,222 @@
+	}
+
+	u := b.URL.JoinPath("/build")
+
+	digest := hmac.Sign(tarFileBytes, []byte(b.hmacSecret), sha256.New)
+	req, err := http.NewRequest(http.MethodPost, u.String(), bytes.NewReader(tarFileBytes))
+	if err != nil {
+		return BuildResult{}, err
+	}
+
+	req.Header.Set("X-Build-Signature", "sha256="+hex.EncodeToString(digest))
+	req.Header.Set("Content-Type", "application/octet-stream")
+	req.Header.Set("User-Agent", "openfaas-go-sdk")
+
+	res, err := b.client.Do(req)
+	if err != nil {
+		return BuildResult{}, err
+	}
+
+	result := BuildResult{}
+	if res.Body != nil {
+		defer res.Body.Close()
+
+		data, err := io.ReadAll(res.Body)
+		if err != nil {
+			return BuildResult{}, err
+		}
+		if err := json.Unmarshal(data, &result); err != nil {
+			return BuildResult{}, err
+		}
+	}
+
+	if res.StatusCode != http.StatusOK && res.StatusCode != http.StatusAccepted {
+		return result, fmt.Errorf("failed to build function, builder responded with status code %d, build status: %s", res.StatusCode, result.Status)
+	}
+
+	return result, nil
+}
+
+// MakeTar create a tar archive that contains the build config and build context.
+func MakeTar(tarPath string, context string, buildConfig *BuildConfig) error {
+	tarFile, err := os.Create(tarPath)
+	if err != nil {
+		return err
+	}
+
+	tarWriter := tar.NewWriter(tarFile)
+	defer tarWriter.Close()
+
+	if err = filepath.Walk(context, func(path string, f os.FileInfo, pathErr error) error {
+		if pathErr != nil {
+			return pathErr
+		}
+
+		targetFile, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+
+		header, err := tar.FileInfoHeader(f, f.Name())
+		if err != nil {
+			return err
+		}
+
+		header.Name = filepath.Join("context", strings.TrimPrefix(path, context))
+		header.Name = strings.TrimPrefix(header.Name, "/")
+
+		if err := tarWriter.WriteHeader(header); err != nil {
+			return err
+		}
+
+		if f.Mode().IsDir() {
+			return nil
+		}
+
+		_, err = io.Copy(tarWriter, targetFile)
+		return err
+	}); err != nil {
+		return err
+	}
+
+	configBytes, err := json.Marshal(buildConfig)
+	if err != nil {
+		return err
+	}
+
+	configHeader := &tar.Header{
+		Name: BuilderConfigFileName,
+		Mode: 0664,
+		Size: int64(len(configBytes)),
+	}
+
+	if err := tarWriter.WriteHeader(configHeader); err != nil {
+		return err
+	}
+
+	if _, err := tarWriter.Write(configBytes); err != nil {
+		return err
+	}
+
+	return err
+}
+
+// Shrinkwrap can be used to create the function build context.
+//
+// The function build context is a normal docker build context. Any valid folder with a Dockerfile will work.
+// `faas-cli build --shrinkwrap` can help here, and allows you to use the existing templates we provide, or one of your own.
+// The Shrinkwrap helper function can be used to create a build. It executes the faas-cli.
+// The faas-cli binary needs to be installed as a separate runtime dependency to use the Shrinkwrap function.
+func Shrinkwrap(image, name string, handler, lang string) error {
```

Comment:
On consideration I think it may be surprising that an SDK requires a binary in order to prepare files for a build.

The CreateBuildContext method in the faas-cli can be extracted, moved here and vendored back.

Add a note about the templates needing to be in place ahead of time when using the Go SDK, explain users can manage templates however they want, including exec'ing the CLI independently.